src/libcore/task/wake.rs

```diff
@@ -4,281 +4,129 @@
-    ///
-    /// Note that most code will not need to call this. Implementers of the
-    /// `UnsafeWake` trait will typically provide a wrapper that calls this
-    /// but you otherwise shouldn't call it directly.
-    ///
-    /// If you're working with the standard library then it's recommended to
-    /// use the `local_waker_from_nonlocal` or `local_waker` to convert a `Waker`
-    /// into a `LocalWaker`.
-    ///
-    /// For this function to be used safely, it must be sound to call `inner.wake_local()`
-    /// on the current thread.
-    #[inline]
-    pub unsafe fn new(inner: NonNull<dyn UnsafeWake>) -> Self {
-        LocalWaker(Waker::new(inner))
-    }
-
-    /// Borrows this `LocalWaker` as a `Waker`.
-    ///
-    /// `Waker` is nearly identical to `LocalWaker`, but is threadsafe
-    /// (implements `Send` and `Sync`).
-    #[inline]
-    pub fn as_waker(&self) -> &Waker {
-        &self.0
-    }
-
-    /// Converts this `LocalWaker` into a `Waker`.
-    ///
-    /// `Waker` is nearly identical to `LocalWaker`, but is threadsafe
-    /// (implements `Send` and `Sync`).
-    #[inline]
-    pub fn into_waker(self) -> Waker {
-        self.0
-    }
-
-    /// Wake up the task associated with this `LocalWaker`.
-    #[inline]
-    pub fn wake(&self) {
-        unsafe { self.0.inner.as_ref().wake_local() }
-    }
-
-    /// Returns whether or not this `LocalWaker` and `other` `LocalWaker` awaken the same task.
-    ///
-    /// This function works on a best-effort basis, and may return false even
-    /// when the `LocalWaker`s would awaken the same task. However, if this function
-    /// returns true, it is guaranteed that the `LocalWaker`s will awaken the same
-    /// task.
-    ///
-    /// This function is primarily used for optimization purposes.
-    #[inline]
-    pub fn will_wake(&self, other: &LocalWaker) -> bool {
-        self.0.will_wake(&other.0)
-    }
-
-    /// Returns whether or not this `LocalWaker` and `other` `Waker` awaken the same task.
-    ///
-    /// This function works on a best-effort basis, and may return false even
-    /// when the `Waker`s would awaken the same task. However, if this function
-    /// returns true, it is guaranteed that the `LocalWaker`s will awaken the same
-    /// task.
-    ///
-    /// This function is primarily used for optimization purposes.
-    #[inline]
-    pub fn will_wake_nonlocal(&self, other: &Waker) -> bool {
-        self.0.will_wake(other)
-    }
-}
-
-impl From<LocalWaker> for Waker {
-    /// Converts a `LocalWaker` into a `Waker`.
-    ///
-    /// This conversion turns a `!Sync` `LocalWaker` into a `Sync` `Waker`, allowing a wakeup
-    /// object to be sent to another thread, but giving up its ability to do specialized
-    /// thread-local wakeup behavior.
-    #[inline]
-    fn from(local_waker: LocalWaker) -> Self {
-        local_waker.0
-    }
-}
-
-impl fmt::Debug for LocalWaker {
+impl fmt::Debug for Waker {
```

Comment:
This would also be more useful by just `#[derive(Debug)]`

Reply:
I thought about it, but it fear this will end up super noisy. I think lots of user-structs will store `Waker` in them and `#[derive(Debug)]`. Those would then all the show 5 different pointers that have no meaning for the average user. 

The main reason why I think it might be useful is that users don't have access to `RawWaker`, so they can't get the inner information.

Maybe we should then only print the `data` and `vtable` pointer for `RawWaker`, and not the inner function pointers. I think if I would need to ever debug those things, the main thing I would be if if `Waker`s refer to the same task or not, and with a lower probability of they use the same vtable. The function pointers are are mostly worthless without names.

In general those things might also be more ones that one would investigate with a real in-memory-debugger than with print debugging.

Reply:
I assume adding something like `debug: Option<unsafe fn(*const (), f: &mut fmt::Formatter) -> fmt::Result>` to `RawWakerVTable` wouldn't be worth it? (and if `RawWakerVTable` is made `non_exhaustive` could be added later anyway).

Reply:
IMO printing the address of the data and vtable pointers seems helpful, but I would leave the substructure out.

Reply:
@cramertj's suggestion seems like a good have-your-🍰-eat-it approach.

Reply:
I added pointer printing for `Waker`. And yes, I don't think the fmt method is worth it.